p2p/peerexchange/discovery.go

```diff
@@ -75,32 +76,59 @@ func New(logger log.Log, h host.Host, config Config) (*Discovery, error) {
 		}
 		bootnodes = append(bootnodes, info)
 	}
+
+	var advertise ma.Multiaddr
+	if len(config.AdvertiseAddress) > 0 {
+		var err error
+		advertise, err = ma.NewMultiaddr(config.AdvertiseAddress)
+		if err != nil {
+			return nil, fmt.Errorf("address to advertise (%s) is invalid: %w", config.AdvertiseAddress, err)
+		}
+		for _, proto := range advertise.Protocols() {
+			if proto.Code == ma.P_P2P {
+				return nil, fmt.Errorf("address to advertise (%s) includes p2p identity", advertise.String())
+			}
+		}
+	} else {
+		var err error
+		advertise, err = ma.NewComponent("tcp", strconv.Itoa(int(portFromHost(logger, h))))
```

Comment:
any particular reason as to why `portFromHost` returns an `uint16` instead of just an `int`? Everywhere it is used the return value is cast to `int` anyway.